### ISSUE TITLE
tests/test_freeimage.py::test_exr_write failure with 2.22.1 release

### ISSUE THREAD

**Issue by mgorny** [imageio/imageio#890]
While testing the 2.22.1 release, I'm hitting the following test failure:

```
FAILED tests/test_freeimage.py::test_exr_write - FileNotFoundError: [Errno 2] No such file or directory: '/tmp/portage/dev-python/im...
```

Apparently, the actual error is (full log below):
```
E               TypeError: Cannot save image of this format to this file type
```

However, the library seems to have OpenEXR support enabled:
```
$ readelf -d /usr/lib64/libfreeimage.so | grep -i exr
 0x0000000000000001 (NEEDED)             Shared library: [libOpenEXR-3_1.so.30]
```

Any suggestion what I could be doing wrong?

Full log with verbose tracebacks: 
[dev-python:imageio-2.22.1:20221003-191506.log](https://github.com/imageio/imageio/files/9700318/dev-python.imageio-2.22.1.20221003-191506.log)

**Comment by FirefoxMetzger** [comment]
That's pretty interesting. I was under the impression that all the FreeImage binaries we use (on the various OSes) are statically linked against their dependencies. You are grabbing yours from `/usr/lib64`, which looks like the system's default.

Are you doing anything specific/custom to link your own version of FreeImage? Also, which OS and FreeImage version are you using?

**Comment by FirefoxMetzger** [comment]
Ah, sorry I think I may have miscommunicated there. I was wondering if you are downloading the library as usual via `imageio.plugins.freeimage.download()` or if you are taking special steps to make ImageIO use the system's version of FreeImage.

**Comment by mgorny** [comment]
Well, I wasn't doing anything special, so… I have no clue. I wouldn't have even imagined imageio would be doing that. Now I see we've got a serious vulnerability on our hands. I'll get back to you after I get this settled first.

**Comment by mgorny** [comment]
I'm pretty sure it uses the system version, though I don't think we've taken any special steps here. It seems to just have worked like this.

**Comment by FirefoxMetzger** [comment]
Interesting. The FreeImage plugin we have is essentially a set of `ctype` bindings that make calls into the FreeImage dll. The code for finding and loading the correct library at runtime lives here (and comes from a time before I was maintainer):

https://github.com/imageio/imageio/blob/eeafb8d5e6b4ec351afa13a9ca05297cd7c8728a/imageio/plugins/_freeimage.py#L483-L515

It eventually calls [`generate_candidate_libs`](https://github.com/imageio/imageio/blob/eeafb8d5e6b4ec351afa13a9ca05297cd7c8728a/imageio/core/findlib.py#L32) which includes a fallback to system libraries if we can't find our local one.

Looking at that code, I think that we can do better in how we go about loading FreeImage. Although my forte isn't in security, so there are others that are more qualified in suggesting improvements.

---

Since this is related, I'll also point out that I am working on https://github.com/imageio/imageio-freeimage on the side. The idea is to eventually move the FreeImage code out of ImageIO itself and make it an optional plugin that people can install if they want to or need to. It will also ship its own version of FreeImage as part of the wheel (and build it in CD), which should be much better from a security standpoint and should allow us to stop searching for a DLL at runtime.

**Comment by mgorny** [comment]
Do I understand correctly that you think that the issue is in freeimage? Ideally I'd try some minimal C reproducer for this but I can't really work on this right now.

That said, you may want to reconsider before spending more effort on freeimage — this library is effectively dead, with last release in 2018, bundling multiple vulnerable libraries. Maintaining a standalone build would be a nightmare. Distros only manage because they remove all bundled libraries and use system libraries but we'd really be happy if we were able to remove freeimage entirely.

**Comment by FirefoxMetzger** [comment]
> Do I understand correctly that you think that the issue is in freeimage?

Yes and no. The test itself attempts to write the image into a `bytes` object. FreeImage doesn't like that so our plugin jumps through an internal hoop of asking FreeImage to write to a temporary file (`/tmp/portage/dev-python/im...`), reads from that temporary location, and then returns the result as `bytes`. 

What is failing is that for some reason that file is not created by FreeImage and then the test fails with the message that you are seeing.

The same process works on GH Actions and on my local machine, so I think it is either the FreeImage version used for the tests or a permissions problem.

> but we'd really be happy if we were able to remove freeimage entirely

Ditto. We mostly maintain the plugin for backward compatibility. Ideally, I would like for this to not happen in the main repo though, which is why I am working on refactoring the code into its own place, where it can exist as an optional dependency for people that still need it (there are still some formats that ImageIO can only read/write via FreeImage for now).

Being in its own repo, however, also allows us to refactor the "download the binary from a GitHub repo" part (which I dislike) and instead use pip to download the binary as part of that optional package, which is arguably more secure than our current homebrew. Further, it might not be wise to ask people to download a binary that is hard to inspect, so building it as part of CD when building the wheel seems much better than treating it as some binary artifact because it allows users to introspect the process.

Does this align with your thinking, or am I misunderstanding you?

**Comment by anjos** [comment]
Also observing this on conda-forge: https://github.com/conda-forge/imageio-feedstock/pull/53, [build logs here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=578255&view=logs&jobId=656edd35-690f-5c53-9ba3-09c10d0bea97&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=e5c8ab1d-8ff9-5cae-b332-e15ae582ed2d)

**Comment by anjos** [comment]
FYI, this continue to be the case on conda-forge with release v2.22.2.  Now there are [further tests failing as well](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=589887&view=logs&jobId=656edd35-690f-5c53-9ba3-09c10d0bea97&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=e5c8ab1d-8ff9-5cae-b332-e15ae582ed2d0).

**Comment by FirefoxMetzger** [comment]
@anjos The two new tests that break are searching for imageio-ffmpeg but not finding it. Indeed, when I checked the list of packages that are installed in the environment (line 432 in the logs) imageio-ffmpeg is not listed there. Further, the logs don't mention any `test_ffmpeg` tests so I assume they are skipped.

Could you check if `imageio-ffmpeg` is actually installed when testing the library?

**Comment by anjos** [comment]
(I don't want to hijack this bug report, and the errors here seem unrelated to the subject.  May be we should move these elsewhere?)

If I enable that module, then the other errors change:

[Without imageio-ffmpeg](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=589887&view=logs&jobId=656edd35-690f-5c53-9ba3-09c10d0bea97&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=e5c8ab1d-8ff9-5cae-b332-e15ae582ed2d0):

```
FAILED tests/test_format.py::test_preferring_arbitrary - AssertionError: asse...
FAILED tests/test_freeimage.py::test_exr_write - FileNotFoundError: [Errno 2]...
ERROR tests/test_format.py::test_missing_format - KeyError: 'imageio.plugins....
```

[With imageio-ffmpeg](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=596240&view=logs&jobId=656edd35-690f-5c53-9ba3-09c10d0bea97&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=e5c8ab1d-8ff9-5cae-b332-e15ae582ed2d):

```
FAILED tests/test_ffmpeg.py::test_writer_pixelformat_size_verbose - AssertionError: assert 'yuv420p' == 'yuv420p(progressive)'
  - yuv420p(progressive)
  + yuv420p
FAILED tests/test_freeimage.py::test_exr_write - FileNotFoundError: [Errno 2] No such file or directory: '/tmp/imageio_tjsep9nm.exr'
FAILED tests/test_pyav.py::test_video_format_to_dtype - ValueError: Can't determine dtype from format `vaapi`. It has no channels.
FAILED tests/test_pyav.py::test_rotation_flag_metadata - KeyError: 'rotate'
```

**Comment by FirefoxMetzger** [comment]
@anjos Yes, this does indeed look like a separate issue from the one discussed here. We should deal with this in a new issue :)
